<a href="https://colab.research.google.com/github/Houssem23x/Janatahack-Healthcare-Analytics/blob/master/Janatahack_Healthcare_Analytics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **1. Prepare Problem**


# a) Load Libraries


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 
from sklearn.model_selection import *
from sklearn.feature_selection import *
from sklearn.model_selection import *
from sklearn.preprocessing import *
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC



# b) Load Datasets



In [ ]:
train = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/Train.csv')
test = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/test_l0Auv8Q.csv')
patient_profile = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/Patient_Profile.csv')
health_camps = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/Health_Camp_Detail.csv')
first_camp = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/First_Health_Camp_Attended.csv')
second_cammp = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/Second_Health_Camp_Attended.csv')
third_camp = pd.read_csv('/content/drive/My Drive/Projects/Janatahack: Healthcare Analytics/Train/Third_Health_Camp_Attended.csv')


# **2.Explore Datasets**


In [ ]:
#Explore the train sataset 
train.head()

Patient_ID  Health_Camp_ID Registration_Date  Var1  Var2  Var3  Var4  Var5
0      489652            6578         10-Sep-05     4     0     0     0     2
1      507246            6578         18-Aug-05    45     5     0     0     7
2      523729            6534         29-Apr-06     0     0     0     0     0
3      524931            6535         07-Feb-04     0     0     0     0     0
4      521364            6529         28-Feb-06    15     1     0     0     7

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75278 entries, 0 to 75277
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         75278 non-null  int64 
 1   Health_Camp_ID     75278 non-null  int64 
 2   Registration_Date  74944 non-null  object
 3   Var1               75278 non-null  int64 
 4   Var2               75278 non-null  int64 
 5   Var3               75278 non-null  int64 
 6   Var4               75278 non-null  int64 
 7   Var5               75278 non-null  int64 
dtypes: int64(7), object(1)
memory usage: 4.6+ MB


* we have some missing values on Registration_Date 
* there are 75278 observations 
* we need to creat the target variables 


In [ ]:
#explore the test dataset
test.head()

Patient_ID  Health_Camp_ID Registration_Date  Var1  Var2  Var3  Var4  Var5
0      505701            6548         21-May-06     1     0     0     0     2
1      500633            6584         02-Jun-06     0     0     0     0     0
2      506945            6582         10-Aug-06     0     0     0     0     0
3      497447            6551         27-Aug-06     0     0     0     0     0
4      496446            6533         19-Sep-06     0     0     0     0     0

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35249 entries, 0 to 35248
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         35249 non-null  int64 
 1   Health_Camp_ID     35249 non-null  int64 
 2   Registration_Date  35249 non-null  object
 3   Var1               35249 non-null  int64 
 4   Var2               35249 non-null  int64 
 5   Var3               35249 non-null  int64 
 6   Var4               35249 non-null  int64 
 7   Var5               35249 non-null  int64 
dtypes: int64(7), object(1)
memory usage: 2.2+ MB


* For now there is no missing values on the test dataset 

In [ ]:
#The patient profile dataset
patient_profile.head(10)

Patient_ID  Online_Follower  ...  City_Type  Employer_Category
0      516956                0  ...        NaN  Software Industry
1      507733                0  ...          H  Software Industry
2      508307                0  ...          D               BFSI
3      512612                0  ...          D          Education
4      521075                0  ...          H             Others
5      502103                0  ...        NaN                NaN
6      528501                0  ...        NaN                NaN
7      519202                0  ...        NaN                NaN
8      521359                0  ...          F                NaN
9      487977                0  ...          G         Technology

[10 rows x 11 columns]

In [ ]:
patient_profile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37633 entries, 0 to 37632
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         37633 non-null  int64 
 1   Online_Follower    37633 non-null  int64 
 2   LinkedIn_Shared    37633 non-null  int64 
 3   Twitter_Shared     37633 non-null  int64 
 4   Facebook_Shared    37633 non-null  int64 
 5   Income             37633 non-null  object
 6   Education_Score    37633 non-null  object
 7   Age                37633 non-null  object
 8   First_Interaction  37633 non-null  object
 9   City_Type          14249 non-null  object
 10  Employer_Category  2840 non-null   object
dtypes: int64(5), object(6)
memory usage: 3.2+ MB


* There are 37633 patients
* the data type of Income,Age and Education_score should be int64 instead of object



In [ ]:
patient_profile['Age'].unique()

array(['39', '40', '46', '47', '80', 'None', '58', '48', '42', '52', '56',
       '49', '38', '45', '51', '50', '43', '53', '41', '44', '66', '37',
       '70', '73', '72', '55', '71', '63', '54', '61', '36', '59', '34',
       '62', '64', '60', '35', '57', '31', '74', '69', '33', '65', '77',
       '75', '67', '76', '68', '32', '78', '30'], dtype=object)

In [ ]:
patient_profile[patient_profile['Age']=='None']['Age'].count()

32618

In [ ]:
patient_profile['Income'].unique()

array(['1', '3', 'None', '4', '0', '2', '6', '5'], dtype=object)

In [ ]:
patient_profile[patient_profile['Income']=='None']['Income'].count()

33271

In [ ]:
patient_profile[patient_profile['Education_Score']=='None']['Education_Score'].count()

36249

* There is 32618 missing values on Age, 33271 on Income and 36249 on Education_Score 
* We have many missing values on City_type and Emploter_Category too
* So, they will not help us in making prediction 

In [ ]:
#The health camp details 
health_camps.head(10)

Health_Camp_ID Camp_Start_Date Camp_End_Date Category1 Category2  Category3
0            6560       16-Aug-03     20-Aug-03     First         B          2
1            6530       16-Aug-03     28-Oct-03     First         C          2
2            6544       03-Nov-03     15-Nov-03     First         F          1
3            6585       22-Nov-03     05-Dec-03     First         E          2
4            6561       30-Nov-03     18-Dec-03     First         E          1
5            6581       07-Dec-03     13-Jun-04     First         F          2
6            6564       17-Dec-03     11-Jun-04     First         C          2
7            6557       04-Jan-04     09-Jan-04     First         C          2
8            6538       09-Jan-04     04-Feb-05     First         F          2
9            6546       09-Jan-04     17-Jan-04     First         E          2

In [ ]:
health_camps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65 entries, 0 to 64
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Health_Camp_ID   65 non-null     int64 
 1   Camp_Start_Date  65 non-null     object
 2   Camp_End_Date    65 non-null     object
 3   Category1        65 non-null     object
 4   Category2        65 non-null     object
 5   Category3        65 non-null     int64 
dtypes: int64(2), object(4)
memory usage: 3.2+ KB


In [ ]:
#The first camp
first_camp.head(10)

Patient_ID  Health_Camp_ID  Donation  Health_Score  Unnamed: 4
0      506181            6560        40      0.439024         NaN
1      494977            6560        20      0.097561         NaN
2      518680            6560        10      0.048780         NaN
3      509916            6560        30      0.634146         NaN
4      488006            6560        20      0.024390         NaN
5      492080            6560        40      0.658537         NaN
6      521555            6560        30      0.536585         NaN
7      493258            6560        30      0.121951         NaN
8      515459            6560        20      0.463415         NaN
9      504261            6560        30      0.878049         NaN

In [ ]:
first_camp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6218 entries, 0 to 6217
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Patient_ID      6218 non-null   int64  
 1   Health_Camp_ID  6218 non-null   int64  
 2   Donation        6218 non-null   int64  
 3   Health_Score    6218 non-null   float64
 4   Unnamed: 4      0 non-null      float64
dtypes: float64(2), int64(3)
memory usage: 243.0 KB


In [ ]:
#The second camp
second_cammp.head(10)

Patient_ID  Health_Camp_ID  Health Score
0      526631            6536      0.875136
1      509122            6536      0.755700
2      498864            6536      0.673181
3      515398            6536      0.722041
4      504624            6536      0.464712
5      486444            6536      0.587405
6      513612            6536      0.722041
7      493571            6536      0.825190
8      494559            6536      0.963084
9      490128            6536      0.755700

In [ ]:
second_cammp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7819 entries, 0 to 7818
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Patient_ID      7819 non-null   int64  
 1   Health_Camp_ID  7819 non-null   int64  
 2   Health Score    7819 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 183.4 KB


In [ ]:
#The third camp
third_camp.head(10)

Patient_ID  ...  Last_Stall_Visited_Number
0      517875  ...                          1
1      504692  ...                          1
2      504692  ...                          1
3      493167  ...                          4
4      510954  ...                          2
5      501825  ...                          4
6      495620  ...                          1
7      526542  ...                          2
8      517351  ...                          5
9      517351  ...                          1

[10 rows x 4 columns]

In [ ]:
third_camp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6515 entries, 0 to 6514
Data columns (total 4 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Patient_ID                 6515 non-null   int64
 1   Health_Camp_ID             6515 non-null   int64
 2   Number_of_stall_visited    6515 non-null   int64
 3   Last_Stall_Visited_Number  6515 non-null   int64
dtypes: int64(4)
memory usage: 203.7 KB


* For the Health_camps and all the three camps  datasets we don't have any missing values 

# **3.Prepare Data**

# a)Data Transfroms

# a) 1. Join Datasets

In [ ]:
#Joining the the train data with health_camps dataset
train_data = pd.merge(train,health_camps,on='Health_Camp_ID')
train_data = pd.merge(train_data,patient_profile,on='Patient_ID')
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 75278 entries, 0 to 75277
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         75278 non-null  int64 
 1   Health_Camp_ID     75278 non-null  int64 
 2   Registration_Date  74944 non-null  object
 3   Var1               75278 non-null  int64 
 4   Var2               75278 non-null  int64 
 5   Var3               75278 non-null  int64 
 6   Var4               75278 non-null  int64 
 7   Var5               75278 non-null  int64 
 8   Camp_Start_Date    75278 non-null  object
 9   Camp_End_Date      75278 non-null  object
 10  Category1          75278 non-null  object
 11  Category2          75278 non-null  object
 12  Category3          75278 non-null  int64 
 13  Online_Follower    75278 non-null  int64 
 14  LinkedIn_Shared    75278 non-null  int64 
 15  Twitter_Shared     75278 non-null  int64 
 16  Facebook_Shared    75278 non-null  int64

In [ ]:
#Joining the test data with health camp datasets 
test_data = pd.merge(test,health_camps,on='Health_Camp_ID')
test_data = pd.merge(test_data,patient_profile,on='Patient_ID')
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 75278 entries, 0 to 75277
Data columns (total 23 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         75278 non-null  int64 
 1   Health_Camp_ID     75278 non-null  int64 
 2   Registration_Date  74944 non-null  object
 3   Var1               75278 non-null  int64 
 4   Var2               75278 non-null  int64 
 5   Var3               75278 non-null  int64 
 6   Var4               75278 non-null  int64 
 7   Var5               75278 non-null  int64 
 8   Camp_Start_Date    75278 non-null  object
 9   Camp_End_Date      75278 non-null  object
 10  Category1          75278 non-null  object
 11  Category2          75278 non-null  object
 12  Category3          75278 non-null  int64 
 13  Online_Follower    75278 non-null  int64 
 14  LinkedIn_Shared    75278 non-null  int64 
 15  Twitter_Shared     75278 non-null  int64 
 16  Facebook_Shared    75278 non-null  int64

# a) 2. Create The Target 

Favorable outcome:
* For the first 2 formats, a favourable outcome is defined as getting a health_score, while in the third format it is defined as visiting at least a stall.
* You need to predict the chances (probability) of having a favourable outcome

In [ ]:
#Patients who have visited at least a stall 
third_camp = third_camp[third_camp.Number_of_stall_visited > 0]


In [ ]:
#Joining the train_data with the three health camps 
train_data = pd.merge(train_data,first_camp,on=['Patient_ID','Health_Camp_ID'],
                      how='left')
train_data = pd.merge(train_data,second_cammp,on=['Patient_ID','Health_Camp_ID'],
                      how='left')
train_data = pd.merge(train_data,third_camp,on=['Patient_ID','Health_Camp_ID'],
                      how='left')
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 75278 entries, 0 to 75277
Data columns (total 29 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Patient_ID                 75278 non-null  int64  
 1   Health_Camp_ID             75278 non-null  int64  
 2   Registration_Date          74944 non-null  object 
 3   Var1                       75278 non-null  int64  
 4   Var2                       75278 non-null  int64  
 5   Var3                       75278 non-null  int64  
 6   Var4                       75278 non-null  int64  
 7   Var5                       75278 non-null  int64  
 8   Camp_Start_Date            75278 non-null  object 
 9   Camp_End_Date              75278 non-null  object 
 10  Category1                  75278 non-null  object 
 11  Category2                  75278 non-null  object 
 12  Category3                  75278 non-null  int64  
 13  Online_Follower            75278 non-null  int

In [ ]:
train_data['Target'] = np.zeros(train_data.shape[0],)

#Condition:
#If one of the this three (first or the second camp health score or the number of stall visited in the third camp)
#is not null the Taget will  take the value 0 otherwise it will take 1
cond = train_data['Health Score'].isna() & train_data['Health_Score'].isna() & train_data['Number_of_stall_visited'].isna()

train_data['Target'] = np.where(cond,train_data['Target'],1)
train_data['Target'].value_counts()


/usr/local/lib/python3.6/dist-packages/pandas/core/computation/expressions.py:178: UserWarning: evaluating in Python space because the '*' operator is not supported by numexpr for the bool dtype, use '&' instead
  f"evaluating in Python space because the {repr(op_str)} "


0.0    54744
1.0    20534
Name: Target, dtype: int64

In [ ]:
train_data.drop(['Health Score','Health_Score','Donation','Unnamed: 4',
                 'Number_of_stall_visited','Last_Stall_Visited_Number'],
                axis=1, inplace = True)
train_data['Target'] = train_data['Target'].astype(int)
train_data

Patient_ID  Health_Camp_ID  ... Employer_Category  Target
0          489652            6578  ...               NaN       1
1          489652            6570  ...               NaN       0
2          489652            6527  ...               NaN       1
3          489652            6539  ...               NaN       0
4          489652            6543  ...               NaN       0
...           ...             ...  ...               ...     ...
75273      511571            6565  ...               NaN       1
75274      489634            6565  ...               NaN       0
75275      497277            6565  ...               NaN       0
75276      492146            6565  ...               NaN       0
75277      487071            6565  ...               NaN       0

[75278 rows x 24 columns]

#a) 3. Dealing with Dates

In [ ]:
train_data.dropna(subset=['Registration_Date'],inplace=True)
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74944 entries, 0 to 75277
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Patient_ID         74944 non-null  int64 
 1   Health_Camp_ID     74944 non-null  int64 
 2   Registration_Date  74944 non-null  object
 3   Var1               74944 non-null  int64 
 4   Var2               74944 non-null  int64 
 5   Var3               74944 non-null  int64 
 6   Var4               74944 non-null  int64 
 7   Var5               74944 non-null  int64 
 8   Camp_Start_Date    74944 non-null  object
 9   Camp_End_Date      74944 non-null  object
 10  Category1          74944 non-null  object
 11  Category2          74944 non-null  object
 12  Category3          74944 non-null  int64 
 13  Online_Follower    74944 non-null  int64 
 14  LinkedIn_Shared    74944 non-null  int64 
 15  Twitter_Shared     74944 non-null  int64 
 16  Facebook_Shared    74944 non-null  int64

In [ ]:
#function that extracts year and month to create columns in data frame 
def extract_date(df,column):
  df[column.replace('_Date','_Year')] = pd.to_datetime(df[column]).apply(lambda x: x.year).astype(int)
  df[column.replace('_Date','_Month')] = pd.to_datetime(df[column]).apply(lambda x: x.month).astype(int)
  df[column.replace('_Date','_Day')] = pd.to_datetime(df[column]).apply(lambda x: x.day).astype(int)
  


In [ ]:
extract_date(train_data,'Registration_Date')
extract_date(train_data,'Camp_Start_Date')
extract_date(train_data,'Camp_End_Date')
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 74944 entries, 0 to 75277
Data columns (total 33 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Patient_ID          74944 non-null  int64 
 1   Health_Camp_ID      74944 non-null  int64 
 2   Registration_Date   74944 non-null  object
 3   Var1                74944 non-null  int64 
 4   Var2                74944 non-null  int64 
 5   Var3                74944 non-null  int64 
 6   Var4                74944 non-null  int64 
 7   Var5                74944 non-null  int64 
 8   Camp_Start_Date     74944 non-null  object
 9   Camp_End_Date       74944 non-null  object
 10  Category1           74944 non-null  object
 11  Category2           74944 non-null  object
 12  Category3           74944 non-null  int64 
 13  Online_Follower     74944 non-null  int64 
 14  LinkedIn_Shared     74944 non-null  int64 
 15  Twitter_Shared      74944 non-null  int64 
 16  Facebook_Shared     74

In [ ]:
extract_date(test_data,'Camp_Start_Date')
extract_date(test_data,'Camp_End_Date')
extract_date(test_data,'Registration_Date')
test_data.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 35249 entries, 0 to 35248
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Patient_ID          35249 non-null  int64 
 1   Health_Camp_ID      35249 non-null  int64 
 2   Registration_Date   35249 non-null  object
 3   Var1                35249 non-null  int64 
 4   Var2                35249 non-null  int64 
 5   Var3                35249 non-null  int64 
 6   Var4                35249 non-null  int64 
 7   Var5                35249 non-null  int64 
 8   Camp_Start_Date     35249 non-null  object
 9   Camp_End_Date       35249 non-null  object
 10  Category1           35249 non-null  object
 11  Category2           35249 non-null  object
 12  Category3           35249 non-null  int64 
 13  Online_Follower     35249 non-null  int64 
 14  LinkedIn_Shared     35249 non-null  int64 
 15  Twitter_Shared      35249 non-null  int64 
 16  Facebook_Shared     35

In [ ]:
train_data['Social_Media'] = train_data['LinkedIn_Shared'] + train_data['Online_Follower'] + train_data['Facebook_Shared'] +train_data['Twitter_Shared']
test_data['Social_Media'] = test_data['LinkedIn_Shared'] + test_data['Online_Follower'] + test_data['Facebook_Shared'] +test_data['Twitter_Shared']

In [ ]:
categorical_features = ['Category1','Category2']
numerical_features = ['Var1','Var2','Var3','Var4','Var5','Category3',
                      'Camp_Start_Month','Camp_End_Month','Registration_Month',
                      'Camp_Start_Year','Camp_End_Year', 'Registration_Year',
                      'Camp_Start_Day','Camp_End_Day', 'Registration_Day',
                      'Social_Media']

In [ ]:
encoder = LabelEncoder()
for col in categorical_features:
  train_data[col] = encoder.fit_transform(train_data[col])
  test_data[col] = encoder.transform(test_data[col])

In [ ]:
features = categorical_features + numerical_features
X = train_data[features]
y = train_data['Target']


In [ ]:
best_features = SelectKBest(score_func=chi2,k='all')
best_features.fit(X,y)
X = best_features.transform(X)
result = dict(zip(features,best_features.scores_))
result_sr = pd.Series(result)
result_sr.sort_values()


Category3                 0.001408
Camp_End_Year             0.365272
Registration_Year         0.767989
Camp_Start_Year           0.867961
Var3                     15.964632
Var4                    140.635793
Camp_Start_Month        326.357228
Camp_Start_Day          341.850187
Registration_Day        356.440283
Social_Media            603.431547
Registration_Month      666.029737
Camp_End_Month          709.145241
Category2               791.631080
Var5                   1842.074221
Var2                   4986.860585
Camp_End_Day           5781.278689
Var1                  11271.455372
Category1             17316.318279
dtype: float64

In [ ]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=0)

In [ ]:
"""
models = []
models.append(('LR',LogisticRegression()))
models.append(('LDA',LinearDiscriminantAnalysis()))
models.append(('KNN',KNeighborsClassifier()))
models.append(('CART',DecisionTreeClassifier()))
models.append(('NB',GaussianNB()))
models.append(('SVM',SVC()))

names = []
results = []
for name,model in models:
  kfold = KFold(n_splits = 5,random_state=0)
  cv_results = cross_val_score(model,X_train,y_train,cv=kfold)
  results.append(cv_results)
  names.append(name)
"""

"\nmodels = []\nmodels.append(('LR',LogisticRegression()))\nmodels.append(('LDA',LinearDiscriminantAnalysis()))\nmodels.append(('KNN',KNeighborsClassifier()))\nmodels.append(('CART',DecisionTreeClassifier()))\nmodels.append(('NB',GaussianNB()))\nmodels.append(('SVM',SVC()))\n\nnames = []\nresults = []\nfor name,model in models:\n  kfold = KFold(n_splits = 5,random_state=0)\n  cv_results = cross_val_score(model,X_train,y_train,cv=kfold)\n  results.append(cv_results)\n  names.append(name)\n"

In [ ]:
"""#Evaluating the results
for name,cv_results in zip(names,results):
  print("%s: %f (%f)" % (name, cv_results.mean(), cv_results.std()))
"""

'#Evaluating the results\nfor name,cv_results in zip(names,results):\n  print("%s: %f (%f)" % (name, cv_results.mean(), cv_results.std()))\n'

In [ ]:
"""
#Evaluating the best 3 models on the test dataset
LR = LogisticRegression().fit(X_train,y_train)
CART = DecisionTreeClassifier().fit(X_train,y_train)
SVC = SVC().fit(X_train,y_train)
best_models = [('LR',LR),
               ('CART',CART),
               ('SVC',SVC)]
"""


"\n#Evaluating the best 3 models on the test dataset\nLR = LogisticRegression().fit(X_train,y_train)\nCART = DecisionTreeClassifier().fit(X_train,y_train)\nSVC = SVC().fit(X_train,y_train)\nbest_models = [('LR',LR),\n               ('CART',CART),\n               ('SVC',SVC)]\n"

In [ ]:
"""
for name,model in best_models:
  print("%s: %f" % (name, model.score(X_test,y_test)))
"""

'\nfor name,model in best_models:\n  print("%s: %f" % (name, model.score(X_test,y_test)))\n'

In [ ]:
"""estimator = KNeighborsClassifier()
params = {'n_neighbors':[1,2,3,4,5,6,7,8,9]}
grid_cv = GridSearchCV(estimator=estimator,param_grid=params)
grid_cv.fit(X_train,y_train)

"""

"estimator = KNeighborsClassifier()\nparams = {'n_neighbors':[1,2,3,4,5,6,7,8,9]}\ngrid_cv = GridSearchCV(estimator=estimator,param_grid=params)\ngrid_cv.fit(X_train,y_train)\n\n"

In [ ]:
"""
print("Best: %f using %s" % (grid_cv.best_score_, grid_cv.best_params_))
means = grid_cv.cv_results_['mean_test_score']
stds = grid_cv.cv_results_['std_test_score']
params = grid_cv.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
  print("%f (%f) with: %r" % (mean, stdev, param))
"""  

'\nprint("Best: %f using %s" % (grid_cv.best_score_, grid_cv.best_params_))\nmeans = grid_cv.cv_results_[\'mean_test_score\']\nstds = grid_cv.cv_results_[\'std_test_score\']\nparams = grid_cv.cv_results_[\'params\']\nfor mean, stdev, param in zip(means, stds, params):\n  print("%f (%f) with: %r" % (mean, stdev, param))\n'

In [ ]:
import xgboost as xgb
model = xgb.XGBClassifier()
model.fit(X_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0,
              learning_rate=0.1, max_delta_step=0, max_depth=3,
              min_child_weight=1, missing=None, n_estimators=100, n_jobs=1,
              nthread=None, objective='binary:logistic', random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
              silent=None, subsample=1, verbosity=1)

In [ ]:
model.score(X_test,y_test)

0.8092601240910001

In [ ]:
from sklearn.ensemble import RandomForestClassifier
RF = RandomForestClassifier(n_estimators=400)
RF.fit(X_train,y_train)
RF.score(X_test,y_test)

0.8074588031222897

In [ ]:
LR = LogisticRegression().fit(X_train,y_train)
LR.score(X_test,y_test)


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


0.7516845686837014

In [ ]:
from sklearn.metrics import roc_auc_score
pred = model.predict(X_test)
roc_auc_score(pred,y_test)

0.762478832603919

In [ ]:
X_sub = test_data[features]
X_sub = best_features.transform(X_sub)
preds = model.predict(X_sub)
preds

array([1, 0, 0, ..., 0, 0, 0], dtype=int64)

In [ ]:
from lightgbm import LGBMClassifier

clf = LGBMClassifier(n_estimators=550,
                     learning_rate=0.03,
                     min_child_samples=40,
                     random_state=1,
                     colsample_bytree=0.5,
                     reg_alpha=2,
                     reg_lambda=2)

clf.fit(X_train, y_train, eval_set=[(X_test, y_test )], verbose=50,
        eval_metric = 'auc', early_stopping_rounds = 100)

preds = clf.predict_proba(X_sub)[:, 1]

Training until validation scores don't improve for 100 rounds.
[50]	valid_0's binary_logloss: 0.445792	valid_0's auc: 0.82726
[100]	valid_0's binary_logloss: 0.43106	valid_0's auc: 0.830619
[150]	valid_0's binary_logloss: 0.42741	valid_0's auc: 0.832638
[200]	valid_0's binary_logloss: 0.42583	valid_0's auc: 0.833718
[250]	valid_0's binary_logloss: 0.424993	valid_0's auc: 0.834318
[300]	valid_0's binary_logloss: 0.424505	valid_0's auc: 0.834731
[350]	valid_0's binary_logloss: 0.424265	valid_0's auc: 0.834893
[400]	valid_0's binary_logloss: 0.424127	valid_0's auc: 0.835013
[450]	valid_0's binary_logloss: 0.424015	valid_0's auc: 0.835106
[500]	valid_0's binary_logloss: 0.423757	valid_0's auc: 0.835335
[550]	valid_0's binary_logloss: 0.423603	valid_0's auc: 0.835466
Did not meet early stopping. Best iteration is:
[547]	valid_0's binary_logloss: 0.423598	valid_0's auc: 0.835472


In [ ]:
output = pd.DataFrame({'Patient_ID':test_data['Patient_ID'],
                        'Health_Camp_ID':test_data['Health_Camp_ID'],
                        'Outcome':preds})
output.to_csv('my_submission.csv', index=False)